### PR TITLE
added story ad badge fix for safari

### PR DESCRIPTION
**Pull request by jshamble**
Fixes the story ad badge for iOS devices by using the static html template in order to load the svg via javascript instead of css. Before this fix, it was pixelated / blurry (on iOS device such as the iphone X), now it is sharp and clear. Removed some unnecessary css rules.

🐛 Bug fix
🖍 CSS / Styling

### CODE REVIEW

File: build-system/eslint-rules/html-template.js [hunk 1]
```diff
@@ -181,7 +181,7 @@ function create(context) {
     const matches = [];
 
     let match;
-    while ((match = invalid.exec(string))) {
+    while ((match = invalid.exec(string)) && !string.includes('<svg')) {
```
Comment (jridgewell) [inline build-system/eslint-rules/html-template.js:184]: Why was this changed?

Comment (powerivq) [inline build-system/eslint-rules/html-template.js:184]: Why not adding `svg` to line 180? That is essentially an allowlist.

**Review by powerivq — CHANGES_REQUESTED** [review]

File: build-system/eslint-rules/html-template.js (hunk 1 above)
Comment (jshamble) [reply build-system/eslint-rules/html-template.js]: Yep, good idea, I have implemented this. Please review again.

Comment (jshamble) [reply build-system/eslint-rules/html-template.js]: I reverted the change and added svgs as exceptions (as they are foreign elements) to the invalid regex line a few lines above. If I don't make this change, the linter will recognize the svg-specific tags as invalid elements,
when they are actually valid.

https://html.spec.whatwg.org/multipage/syntax.html#foreign-elements

File: build-system/eslint-rules/html-template.js [hunk 2]
```diff
@@ -177,7 +177,7 @@ function create(context) {
     // Void tags are defined at
     // https://html.spec.whatwg.org/multipage/syntax.html#void-elements
     const invalid =
-      /<(?!area|base|br|col|embed|hr|img|input|link|meta|param|source|track|wbr)([a-zA-Z-]+)([\s\n][^>]*)?\/>/gm;
+      /<(?!area|base|br|col|embed|hr|img|input|link|meta|param|source|track|wbr)(((svg)?)([a-zA-Z-])+)([\s\n][^>]*)?\/>/gm;
```
Comment (jridgewell) [inline build-system/eslint-rules/html-template.js:180]: It seems `svgSelfClosing` and related code can be cleaned up.

Comment (jridgewell) [inline build-system/eslint-rules/html-template.js:180]: Skipping the call to `invalidVoidTag` for `svg` tags would fix this without requiring full regex scans.

**Review by jridgewell — CHANGES_REQUESTED** [review]

File: build-system/eslint-rules/html-template.js (hunk 2 above)
Comment (jshamble) [reply build-system/eslint-rules/html-template.js:180]: agreed, and changed to skip the call to invalidVoidTag for svg elements.

Comment (jshamble) [reply build-system/eslint-rules/html-template.js]: This could mean a lot of things. Can you clarify what you mean by 'cleaning up'?

Comment (jridgewell) [reply build-system/eslint-rules/html-template.js]: The link you provided says that all foreign elements are valid void tags, so the automatic "fix" isn't necessary at all for SVG elements. Can you delete everything related to that?

File: build-system/eslint-rules/html-template.js [hunk 3]
```diff
@@ -135,7 +135,12 @@ function create(context) {
       });
     }
 
-    const invalids = invalidVoidTag(string);
+    let invalids = 0;
```
Comment (jridgewell) [inline build-system/eslint-rules/html-template.js:138]: `invalids` is supposed to be an array.

File: build-system/eslint-rules/html-template.js (hunk 2 above)
Comment (jshamble) [reply build-system/eslint-rules/html-template.js]: https://html.spec.whatwg.org/multipage/syntax.html#void-elements

If you scroll up just a tad

"There are six different kinds of elements: void elements, the template element, raw text elements, escapable raw text elements, foreign elements, and normal elements."

svg is considered a foreign element, so I don't think we should consider them as void tags.

We have a few options here, but one solution could be:

make invalidVoidTag => invalidTag, and add svg to the list of tags

I think a quick meeting can help us figure this out faster - i've set a few minutes for us to meet tomorrow.

Comment (jridgewell) [reply build-system/eslint-rules/html-template.js]: See https://html.spec.whatwg.org/multipage/syntax.html#syntax-start-tag, Step 6

> Then, if the element is one of the void elements, or if the element is a foreign element, then there may be a single U+002F SOLIDUS character (/). This character has no effect on void elements, but on foreign elements it marks the start tag as self-closing.

Comment (jshamble) [reply build-system/eslint-rules/html-template.js]: @jridgewell just cleaned-up and ran some tests, so we should be good to go -
 give it a look over when you get the chance and let me know if you need anything else.

**Review by jridgewell — APPROVED** [review]

**Review by powerivq — APPROVED** [review]

**Comment by calebcordry** [comment]
I think we accidentally changed the UI a bit here. 

**OLD**

image
image

**NEW**

image
image

**DESIGN SPEC** 
image

Let's get a fix in ASAP. Also we should write a test to prevent similar issues in the future.

**Comment by newmuis** [comment]
Super curious: why does loading the SVG via JS instead of CSS fix this?

**Comment by jridgewell** [comment]
The CSS background is being rasterized once to a PNG (internally in the browser), and then being resized. Creating true SVG nodes allows the browser to scale the SVG infinitely, making the text appear much crisper.
